src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -454,6 +462,46 @@ protected override int AddVariable(VariableIdentifier identifier)
             return _variables.Add(identifier);
         }
 
+        // Note: this is for placeholders to unvisited expressions
+        private void AddPlaceholder(BoundValuePlaceholderBase placeholder, BoundExpression unvisited)
+        {
+            _placeholdersToUnvisitedExpressionOpt ??= PooledDictionary<BoundValuePlaceholderBase, BoundExpression>.GetInstance();
+            _placeholdersToUnvisitedExpressionOpt.Add(placeholder, unvisited);
+        }
+
+        // Note: this is for placeholders to unvisited expressions
+        private void RemovePlaceholder(BoundValuePlaceholderBase placeholder)
+        {
+            Debug.Assert(_placeholdersToUnvisitedExpressionOpt is not null);
+            _placeholdersToUnvisitedExpressionOpt.Remove(placeholder);
+        }
+
+        // Note: this is for placeholders to unvisited expressions
+        private bool TryReplacePlaceholder(BoundValuePlaceholderBase placeholder, [NotNullWhen(true)] out BoundExpression? unvisited)
+        {
+            if (_placeholdersToUnvisitedExpressionOpt is not null &&
+                _placeholdersToUnvisitedExpressionOpt.TryGetValue(placeholder, out unvisited))
+            {
+                return true;
```

Comment:
If we're not revisiting expressions, it seems like we can remove the placeholder from the dictionary here.